src/runtime_src/core/pcie/common/memaccess.h

```diff
@@ -164,37 +164,39 @@ namespace xcldev {
      * Caller's responsibility to do sanity checks. No sanity checks done here
      */
     int readBank(std::ofstream& aOutFile, unsigned long long aStartAddr, unsigned long long aSize) {
-      char *buf = 0;
+      auto buf = std::make_shared<char>();
```

Comment:
This char* wrapped by 'buf' will be deleted using 'delete', but posix_memalign probably expects "free".  While this may work on Linux, it is not correct.  Match new  with delete, and malloc with free.

Also, it doesn't look like this should be shared ownership, looks like 'buf' is really strictly owned within the function scope.

I think we could wrap aligned_alloc (or posix_memalign) to return a unique_ptr.
Here is an example that works.  For shared usage, I think the allocation function should be declared in  memalign.h and defined in memalign.cpp (runtime_src/core/common)
```
#include <memory>
#include <cstdlib>

namespace xrt_core {
static auto deleter = [](void* ptr) { free(ptr); };
std::unique_ptr<void, decltype(deleter)>
aligned_alloc(size_t align, size_t size)
{
  return std::unique_ptr<void, decltype(deleter)>(::aligned_alloc(align, size),deleter);
}
} // namespace xrt_core
```
On success aligned_alloc returns a valid ptr, on failure it returns nullptr.  
Your code would simply call
`auto buf = xrt_core::aligned_alloc(get_pagesize(), blockSize);`

Reply:
Actually on second thought, this aligned alloc would have to be in header file but within namespace.  The type of 'deleter' is needed.

Reply:
How about this
```
% cat src/runtime_src/core/common/memalign.h
#include <memory>
#include <cstdlib>

namespace xrt_core {
...
auto aligned_ptr_deleter = [] (void* ptr) { free(ptr); } ;
using aligned_ptr_type = std::unique_ptr<void, decltype(aligned_ptr_deleter)>;
inline aligned_ptr_type
aligned_alloc(size_t align, size_t size)
{
  return aligned_ptr_type(::aligned_alloc(align, size),aligned_ptr_deleter);
}
} // xrt_core
```